dramatiq/middleware/shutdown.py

```diff
@@ -0,0 +1,83 @@
+# This file is a part of Dramatiq.
+#
+# Copyright (C) 2017,2018 CLEARTYPE SRL <bogdan@cleartype.io>
+#
+# Dramatiq is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or (at
+# your option) any later version.
+#
+# Dramatiq is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+# License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import threading
+import warnings
+
+from ..logging import get_logger
+from .middleware import Middleware
+from .threading import Interrupt, current_platform, supported_platforms, raise_thread_exception
+
+
+class Shutdown(Interrupt):
+    """Exception used to interrupt worker threads when their worker
+    processes have been signaled for termination.
+    """
+
+
+class ShutdownNotifications(Middleware):
+    """Middleware that interrupts actors whose worker process has been
+    signaled for termination.
+    Currently, this is only available on CPython.
+
+    Note:
+      This works by setting an async exception in the worker thread
+      that runs the actor.  This means that the exception will only get
+      called the next time that thread acquires the GIL.  Concretely,
+      this means that this middleware can't cancel system calls.
+
+    Parameters:
+      notify_shutdown(bool): When true, the actor will be interrupted
+        if the worker process was terminated.
+    """
+
+    def __init__(self):
+        self.logger = get_logger(__name__, type(self))
+        self.notifications = set()
+
+    @property
+    def actor_options(self):
+        return {"notify_shutdown"}
+
+    def after_process_boot(self, broker):
+        if current_platform not in supported_platforms:  # pragma: no cover
+            msg = "ShutdownNotifications cannot kill threads on your current platform (%r)."
+            warnings.warn(msg % current_platform, category=RuntimeWarning, stacklevel=2)
+
+    def before_worker_shutdown(self, broker, worker):
+        self.logger.debug("Sending shutdown notification to worker threads...")
+        for thread_id in self.notifications:
+            self.logger.info("Worker shutdown notification. Raising exception in worker thread %r.", thread_id)
```

Comment:
Is it relevant to include the `thread_id` here, since the worker's log format includes the thread name?

Reply:
I believe so, because the thread logging the message may be different from the thread executing the task.